### ISSUE TITLE
Have key_or_default take default if key's value is empty

### ISSUE THREAD

**Issue by smreichling** [hashicorp/consul-template#449]
Would be nice to be able to protect against the case of a blank value under a key in Consul resulting in invalid templating output.

**Comment by sethvargo** [comment]
Hi @smreichling 

I've been thinking about this for some time, and I think this is not the best behavior for Consul Template. Many users utilize the existence of a key in Consul's KV store to control flow or values. If Consul Template treated a key with empty data (`""`) the same way it treats a non-existent key, it would severely limit the flexibility of that switch.

I would love to hear a bit more about your use case and why you think we should treat a key with an empty value the same as a non-existent key.

Please note, you can also use a `len` check in the template to check if the value of the key is nil, for example:

```go
{{ $result := key "foo" }}
{{ if $result }}
  {{ $result }}
{{ end }}
```

**Comment by RyanGordon** [comment]
Hi @sethvargo,

Our suggestion for this feature is mainly birthed from an abundance of caution for operation error. In our specific setup, a blank value on an existent key is still invalid and would benefit from a default-type behavior. I do agree that there are use cases where a blank value would be valid and so I think the suggestion would be better amended to suggest introducing a new function similar to key_or_default, such as key_with_blank_default or whatever could be worded best to describe the difference in the two functionalities cleanly. We could certainly use a larger if statement to achieve the same thing but prefer the nicety of a single, simple line to get and default the value. What do you think of this suggestion?

**Comment by sethvargo** [comment]
Hi @RyanGordon 

I'm rather hesitant to add an additional API, especially since everyone's definition of "blank" is different. Do you consider a key with a value of `" "` (space) blank? This is why I think the check for a non-existent key makes the most sense.

**Comment by RyanGordon** [comment]
Maybe we can be more specific then? Something like: "key_with_null_default"?

**Comment by sethvargo** [comment]
Hi @RyanGordon 

I think the [documentation](https://github.com/hashicorp/consul-template/tree/8a8a9301631244754792e5bf979d28ec6e3a9900#key_or_default) makes this very clear:

> Query Consul for the value at the given key. If no key exists at the given path, the default value will be used instead.

I am also against renaming a function that is already part of the public API for compatibility reasons.

**Comment by RyanGordon** [comment]
Hey @sethvargo,

I wasn't suggesting renaming the current key_or_default function, but rather introducing a new one that could handle the additional specific case of a key existing but with a null (empty) value. So essentially, in pseudo-code I would see it working like this:
```
key_with_null_default defaultarg:
    key = key_or_default defaultarg
    if key:
       return key
    return defaultarg
``` 

This way it would handle both the non-exist case and the exists-but-null case

**Comment by sethvargo** [comment]
@RyanGordon right, but as I said, everyone's definition of "blank" is different. If you want this level of granularity in your templates, I think you should do a `len` check explicitly instead of relying on a function.

**Comment by RyanGordon** [comment]
I understand where you're coming from. I avoided the ambiguous usage of the "blank" word in my previous example and instead used the wording "null" which I think is as specific as we can be for this situation. I updated by pseudo-code to do a len check explicitly to be clearer. I could always just do the same thing in my own template, I just thought it would be worthy of bringing it up for discussion here as a possible candidate function for templates
